scripts/esrupgrades/esr.5.37-7.8.mysql.up.sql

```diff
@@ -1,1295 +1,1376 @@
-    (
-        SELECT Count(*) FROM Information_Schema.Columns
-        WHERE table_name = 'PluginKeyValueStore'
-        AND table_schema = DATABASE()
-        AND column_name = 'PKey'
-        AND column_type != 'varchar(150)'
-    ) > 0,
-    'ALTER TABLE PluginKeyValueStore MODIFY COLUMN PKey varchar(150);',
-    'SELECT 1'
-));
-
-PREPARE alterTypeIfExists FROM @preparedStatement;
-EXECUTE alterTypeIfExists;
-DEALLOCATE PREPARE alterTypeIfExists;
+	-- MODIFY COLUMN Props
+	DECLARE ModifyProps BOOLEAN;
+	DECLARE ModifyPropsQuery TEXT DEFAULT NULL;
 
-/* ==> mysql/000074_upgrade_users_v6.3.up.sql <== */
+	-- 'CREATE INDEX idx_posts_root_id_delete_at ON Posts(RootId, DeleteAt);'
+	DECLARE CreateIndexRootId BOOLEAN;
+	DECLARE CreateIndexRootIdQuery TEXT DEFAULT NULL;
 
-SET @preparedStatement = (SELECT IF(
-    (
-        SELECT COUNT(*) FROM INFORMATION_SCHEMA.COLUMNS
-        WHERE table_name = 'Users'
-        AND table_schema = DATABASE()
-        AND column_name = 'AcceptedTermsOfServiceId'
-    ) > 0,
-    'ALTER TABLE Users DROP COLUMN AcceptedTermsOfServiceId;',
-    'SELECT 1'
-));
-
-PREPARE alterIfExists FROM @preparedStatement;
-EXECUTE alterIfExists;
-DEALLOCATE PREPARE alterIfExists;
+	-- 'DROP INDEX idx_posts_root_id ON Posts;',
+	DECLARE DropIndex BOOLEAN;
+	DECLARE DropIndexQuery TEXT DEFAULT NULL;
 
-/* ==> mysql/000075_alter_upload_sessions_index.up.sql <== */
+	-- 'CREATE INDEX idx_posts_create_at_id on Posts(CreateAt, Id) LOCK=NONE;'
+	DECLARE CreateIndexCreateAt BOOLEAN;
+	DECLARE CreateIndexCreateAtQuery TEXT DEFAULT NULL;
+
+	SELECT COUNT(*) FROM INFORMATION_SCHEMA.COLUMNS
+		WHERE TABLE_NAME = 'Posts'
+		AND table_schema = DATABASE()
+		AND COLUMN_NAME = 'ParentId'
+		INTO DropParentId;
+
+	SELECT COUNT(*) FROM INFORMATION_SCHEMA.COLUMNS
+		WHERE table_name = 'Posts'
+		AND table_schema = DATABASE()
+		AND column_name = 'FileIds'
+		AND LOWER(column_type) != 'text'
+		INTO ModifyFileIds;
+
+	SELECT COUNT(*) FROM INFORMATION_SCHEMA.COLUMNS
+		WHERE table_name = 'Posts'
+		AND table_schema = DATABASE()
+		AND column_name = 'Props'
+		AND LOWER(column_type) != 'json'
+		INTO ModifyProps;
+
+	SELECT COUNT(*) = 0 FROM INFORMATION_SCHEMA.STATISTICS
+		WHERE table_name = 'Posts'
+		AND table_schema = DATABASE()
+		AND index_name = 'idx_posts_root_id_delete_at'
+		INTO CreateIndexRootId;
+
+	SELECT COUNT(*) FROM INFORMATION_SCHEMA.STATISTICS
+		WHERE table_name = 'Posts'
+		AND table_schema = DATABASE()
+		AND index_name = 'idx_posts_root_id'
+		INTO DropIndex;
+
+	SELECT COUNT(*) = 0 FROM INFORMATION_SCHEMA.STATISTICS
+		WHERE table_name = 'Posts'
+		AND table_schema = DATABASE()
+		AND index_name = 'idx_posts_create_at_id'
+		INTO CreateIndexCreateAt;
+
+	IF DropParentId THEN
+		SET DropParentIdQuery = 'DROP COLUMN ParentId';
+		UPDATE Posts SET RootId = ParentId WHERE RootId = '' AND RootId != ParentId;
+	END IF;
+
+	IF ModifyFileIds THEN
+		SET ModifyFileIdsQuery = 'MODIFY COLUMN FileIds text';
+	END IF;
```

Comment:
Would it make sense to structure the procedures as:

```
SELECT ... INTO ConditionalVariable1
IF ConditionalVariable1 THEN
    ...
END IF;

SELECT ... INTO ConditionalVariable2
IF ConditionalVariable2 THEN
    ...
END IF;
```

One thing I'm struggling with is to pair up the checks with the executions.

Reply:
I'm 0/5 on this given it already works, but just to illustrate perhaps what I was thinking -- and throwing in a unified `ShouldApplyChange` variable to illustrate another simplification:

```sql
DELIMITER //
CREATE PROCEDURE MigrateUploadSessions ()
BEGIN
	DECLARE ShouldApplyChange BOOLEAN;

	DECLARE CreateIndexQuery TEXT DEFAULT NULL;
	SELECT COUNT(*) = 0 FROM INFORMATION_SCHEMA.STATISTICS
		WHERE table_name = 'UploadSessions'
		AND table_schema = DATABASE()
		AND index_name = 'idx_uploadsessions_type'
		INTO ShouldApplyChange;

	IF ShouldApplyChange THEN
		SET CreateIndexQuery = 'ADD INDEX idx_uploadsessions_type (Type)';
	END IF;

	DECLARE AlterIndexQuery TEXT DEFAULT NULL;
	SELECT IFNULL(GROUP_CONCAT(column_name ORDER BY seq_in_index), '') = 'Type' FROM information_schema.statistics
		WHERE table_name = 'UploadSessions'
		AND table_schema = DATABASE()
		AND index_name = 'idx_uploadsessions_user_id'
		GROUP BY index_name
		INTO ShouldApplyChange;

	IF ShouldApplyChange THEN
		SET AlterIndexQuery = 'DROP INDEX idx_uploadsessions_user_id, ADD INDEX idx_uploadsessions_user_id (UserId)';
	END IF;

	DECLARE AlterColumnQuery TEXT DEFAULT NULL;
	SELECT COUNT(*) FROM INFORMATION_SCHEMA.COLUMNS
		WHERE table_name = 'UploadSessions'
		AND table_schema = DATABASE()
		AND column_name = 'Type'
		AND REPLACE(LOWER(column_type), '"', "'") != "enum('attachment','import')"
		INTO ShouldApplyChange;
		
	IF ShouldApplyChange THEN
		SET AlterColumnQuery = 'MODIFY COLUMN Type ENUM("attachment", "import")';
	END IF;

        SET @alterQuery = CONCAT_WS(', ', CreateIndexQuery, AlterIndexQuery, AlterColumnQuery)
	IF alterQuery <> '' THEN
		SET @query = CONCAT('ALTER TABLE UploadSessions ', alterQuery);
		PREPARE stmt FROM @query;
		EXECUTE stmt;
		DEALLOCATE PREPARE stmt;
	END IF;
END//
DELIMITER ;
SELECT CONCAT('-- ', NOW(), ' MigrateUploadSessions procedure starting.') AS DEBUG;
CALL MigrateUploadSessions();
SELECT CONCAT('-- ', NOW(), ' MigrateUploadSessions procedure finished.') AS DEBUG;
DROP PROCEDURE IF EXISTS MigrateUploadSessions;
```

Reply:
Sounds good!!